src/libopensc/card-cardos.c

```diff
@@ -525,15 +525,17 @@ static int cardos_list_files(sc_card_t *card, u8 *buf, size_t buflen)
 	while (len != 0) {
 		size_t   tlen = 0, ilen = 0;
 		/* is there a file information block (0x6f) ? */
-		p = sc_asn1_find_tag(card->ctx, p, len, 0x6f, &tlen);
-		if (p == NULL) {
+		tag = sc_asn1_find_tag(card->ctx, p, len, 0x6f, &tlen);
+		if (tag == NULL) {
 			sc_log(card->ctx,  "directory tag missing");
 			return SC_ERROR_INTERNAL;
 		}
+		len -= tlen - (tag - p);
```

Comment:
this is the problem. This lines evaluates to `len = len - (tlen - (tag - p));`, while we want `len = (len - tlen) - (tag - p);`. Fixing this makes the cardos card working here again. I would suggest to rewrite it for readability like this:
```suggestion
		len = len - tlen - (tag - p);
```